File: LeetCode题目集/202. Happy Number(easy).cpp
/*
 * @Author: SourDumplings
 * @Date: 2019-08-09 11:24:05
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/happy-number/
 */

class Solution
{
public:
    bool isHappy(int n)
    {
        int checkNum = 100;
        while (checkNum--)
        {
            if (check(n))
            {
                return true;
            }
        }
        return false;
    }

    bool check(int &n)
    {
        string s = to_string(n);
        int sum = 0;
        for (char d : s)
        {
            int dd = d - '0';
            sum += dd * dd;
        }
        if ((n = sum) == 1)
        {
            return true;
        }
        return false;
    }
};

File: LeetCode题目集/202. Happy Number(easy).java
/*
 * @Author: SourDumplings
 * @Date: 2019-08-09 11:32:24
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/happy-number/
 */

class Solution
{
    public boolean isHappy(int n)
    {
        int repeatNum = 100;
        while (repeatNum-- != 0)
        {
            if ((n = doCheck(n)) == 1)
                return true;

        }
        return false;
    }

    private int doCheck(int n)
    {
        String num = new Integer(n).toString();
        int l = num.length();
        int res = 0;
        for (int i = 0; i < l; i++)
        {
            Character d = num.charAt(i);
            int digit = Integer.parseInt(d.toString());
            res += digit * digit;
        }
        return res;
    }
}